dimod/include/dimod/abc.h

```diff
@@ -0,0 +1,1083 @@
+        return;
+    } else if (source == Vartype::SPIN && target == Vartype::BINARY) {
+        if (has_adj()) {
+            for (auto it = (*adj_ptr_)[v].begin(); it != (*adj_ptr_)[v].end(); ++it) {
+                linear_biases_[it->v] -= it->bias;
+                asym_quadratic(it->v, v) *= 2;  // log(n)
+                it->bias *= 2;
+            }
+        }
+
+        offset_ -= linear_biases_[v];
+        linear_biases_[v] *= 2;
+    } else if (source == Vartype::BINARY && target == Vartype::SPIN) {
+        if (has_adj()) {
+            for (auto it = (*adj_ptr_)[v].begin(); it != (*adj_ptr_)[v].end(); ++it) {
+                linear_biases_[it->v] += it->bias / 2;
+                asym_quadratic(it->v, v) /= 2;  // log(n)
+                it->bias /= 2;
+            }
+        }
+
+        offset_ += linear_biases_[v] / 2;
+        linear_biases_[v] /= 2;
+    } else {
+        // todo: there are more we could support
+        throw std::logic_error("unsupported vartype change");
+    }
+}
+
+template <class bias_type, class index_type>
+void QuadraticModelBase<bias_type, index_type>::change_vartypes(Vartype source, Vartype target) {
+    if (source == target) return;  // nothing to do
+
+    bias_type lin_mp, lin_offset_mp, quad_mp, quad_offset_mp, lin_quad_mp;
+    if (source == Vartype::SPIN && target == Vartype::BINARY) {
+        lin_mp = 2;
+        lin_offset_mp = -1;
+        quad_mp = 4;
+        lin_quad_mp = -2;
+        quad_offset_mp = .5;
+    } else if (source == Vartype::BINARY && target == Vartype::SPIN) {
+        lin_mp = .5;
+        lin_offset_mp = .5;
+        quad_mp = .25;
+        lin_quad_mp = .25;
+        quad_offset_mp = .125;
+    } else {
+        throw std::logic_error("unsupported vartype combo");
+    }
+
+    for (size_type u = 0; u < num_variables(); ++u) {
+        bias_type lbias = linear_biases_[u];
+
+        linear_biases_[u] *= lin_mp;
+        offset_ += lin_offset_mp * lbias;
+
+        if (has_adj()) {
+            for (auto it = (*adj_ptr_)[u].begin(); it != (*adj_ptr_)[u].end(); ++it) {
+                bias_type qbias = it->bias;
+
+                it->bias *= quad_mp;
+                linear_biases_[u] += lin_quad_mp * qbias;
+                offset_ += quad_offset_mp * qbias;
+            }
+        }
+    }
+}
+
+template <class bias_type, class index_type>
+void QuadraticModelBase<bias_type, index_type>::clear() {
+    adj_ptr_.reset(nullptr);
+    linear_biases_.clear();
+    offset_ = 0;
+}
+
+template <class bias_type, class index_type>
+template <class Iter>
+bias_type QuadraticModelBase<bias_type, index_type>::energy(Iter sample_start) const {
+    static_assert(std::is_same<std::random_access_iterator_tag,
+                               typename std::iterator_traits<Iter>::iterator_category>::value, "iterators must be random access");
+
+    bias_type en = offset();
+
+    if (has_adj()) {
+        for (index_type u = 0; static_cast<size_type>(u) < num_variables(); ++u) {
```

Comment:
This form of energy calculation is slightly less accurate than the initial calculation in the c++ version where simd code is used and in essence multiple accumulators are used. If this needs to be the most accurate version then we need to use simd code or at least have multiple accumulators or apply a compensation based algorithm like the Kahan sum.

But then again, its probably okay as it is :D

Reply:
Agree, let's leave it alone for right now.